crowbar_framework/app/models/service_object.rb

```diff
@@ -1725,4 +1741,38 @@ def active_update(proposal, inst, in_queue, bootstrap = false)
       [400, e2.message]
     end
   end
+
+  def skip_unready_nodes(bc, inst, new_elements, old_elements)
+    skip_unready_nodes_config = Rails.application.config.experimental["skip_unready_nodes"]
+    # return if the feature is disabled
+    return new_elements unless skip_unready_nodes_config["enabled"]
+    logger.warn("Experimental skip_unready_nodes feature enabled")
+    logger.debug("skip_unready_nodes: enter for #{bc}:#{inst}")
+    nodes_pre_cached = {}
+    cleaned_elements = new_elements.deep_dup
+    skip_unready_nodes_config["roles"].each do |role|
+      # only do something if we have the same role on both old and new
+      next unless new_elements.key?(role) || old_elements.key?(role)
+      # we only can skip nodes that are on both old and new, as we know that those old nodes had
+      # the roles applied and will eventually become consistent with the deployment due to the
+      # periodic chef run
+      shared_elements = new_elements[role] & old_elements[role]
+      shared_elements.each do |n|
+        node = Node.find_by_name(n)
+        next if node.nil?
+        nodes_pre_cached[n] = node
+        # skip if nodes are on ready or crowbar_upgrade state, we dont need to do anything
+        next if ["ready", "crowbar_upgrade"].include?(node.crowbar["state"])
+        logger.warn(
+          "Node #{n} is being removed from the deployment for #{bc}:#{inst} with role #{role}"
+        )
+        logger.warn(
+          "Node #{n} state is: #{node.crowbar["state"]}"
+        )
```

Comment:
I really dislike that there are two log lines and not just one, but if you care about it, then keep it that way.

Reply:
I think the second line can be removed, I just used as a debug flag during tests to make sure that the nodes being skipped are really in a non-ready status.